### ISSUE TITLE
rework `feed_union.rb`

### ISSUE THREAD

**Issue by larkost** [rethinkdb/rethinkdb#4314]
In the `raft` branch the `feed_union.rb` test is currently failing in an unhelpful way (basically: "something is timing out waiting for `.next`, but I am not going to tell you where). After working on this for a while I am confused (as is @mlucy) about why it is succeeding in `next`, so feel the need to rewrite it. Having worked to fix the current incarnation for a while yesterday I think that is needs larger change and so am proposing a re-write into Python in order to use the `unitest` facilities that we have there.

Goals of the re-write:
* Be very explicit about what fails
* The tests should be easy to understand
* Break the tests into small units that tests specific aspects
* Expected output should be readable and probably not automatically generated
* Cover `stream`s and `changefeed`s with `limit`
* Include one test with a lot of `union`'ed `changefeeds`

**Comment by larkost** [comment]
A list of tests to help me decide what language to write this in:
- [ ] simple union combinations
  - [ ] three streams
  - [ ] three limited streams
  - [ ] three eager streams (e.g.: `r.expr([1,2,3])`)
  - [ ] three changefeeds (table, point, range)
  - [ ] three limited changefeeds
  - [ ] one each of: simple stream, eager stream, changefeed, limited changefeed
- [ ] nested unions
  - [ ] deeply nested: `[a, [b, [c]]]` (but a dozen or more layers deep)
  - [ ] wide and shallow: `[[a, b, c, d], [e, f, g, h]] (but with a dozen or more in each)
- [ ] unions are terminated if table is dropped
  - [ ] only single table involved
  - [ ] multiple tables, but only one involved

Each of these should be run through following checks:
- initial values
- empty after initial values (e.g.: `next(wait=.2)` times out)
- returns a set of changes (or not)
- errors after table dropped

**Comment by danielmewes** [comment]
The fact that this test might benefit from a rewrite aside, do we actually understand why it is currently failing on `next` (including raft)?

**Comment by larkost** [comment]
No, we never figured out why the timeout is happening. Attempts to reproduce it outside the test were inconclusive.

**Comment by mlucy** [comment]
Moving to the 2.1 milestone, we should make sure to figure this out before we release.  I vaguely remember looking into this before and thinking it might be a server bug.

**Comment by danielmewes** [comment]
Temporarily reassigning to @mlucy to figure out why this fails.

Please assign back to @larkost later for subsequent test improvements.

**Comment by larkost** [comment]
I am taking this back as we have two issues and I am giving @mlucy the other (#4057) to do that work under.

**Comment by danielmewes** [comment]
Ok putting the other one into 2.1 and this one back to `tests` then.
